File: src/main/java/com/saas/pm/entity/Project.java
package com.saas.pm.entity;

import jakarta.persistence.*;
import java.time.LocalDate;
import java.util.List;

/**
 * Project entity represents work initiatives that contain tasks and resources.
 * Each project belongs to a tenant and has a creator/manager.
 * Projects can have multiple team members and track progress through tasks.
 */
@Entity
@Table(name = "projects")
public class Project extends BaseEntity {

    @Column(name = "name", nullable = false)
    private String name;

    @Column(name = "description", columnDefinition = "TEXT")
    private String description;

    @Column(name = "code", nullable = false, unique = true)
    private String code;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false)
    private ProjectStatus status = ProjectStatus.PLANNING;

    @Enumerated(EnumType.STRING)
    @Column(name = "priority", nullable = false)
    private ProjectPriority priority = ProjectPriority.MEDIUM;

    @Column(name = "start_date")
    private LocalDate startDate;

    @Column(name = "end_date")
    private LocalDate endDate;

    @Column(name = "estimated_hours")
    private Integer estimatedHours;

    @Column(name = "actual_hours")
    private Integer actualHours;

    @Column(name = "budget")
    private Double budget;

    @Column(name = "currency", nullable = false)
    private String currency = "USD";

    @Column(name = "is_public", nullable = false)
    private Boolean isPublic = false;

    @Column(name = "tags")
    private String tags;

    @Column(name = "settings", columnDefinition = "TEXT")
    private String settings;

    // Relationships
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "created_by_id", nullable = false)
    private User creator;

    @OneToMany(mappedBy = "project", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    private List<Task> tasks;

    @OneToMany(mappedBy = "project", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    private List<ProjectMember> members;

    @OneToMany(mappedBy = "project", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    private List<FileAttachment> attachments;

    // Constructors
    public Project() {
        super();
    }

    public Project(String name, String code, User creator, Long tenantId) {
        super(tenantId);
        this.name = name;
        this.code = code;
        this.creator = creator;
    }

    // Getters and Setters
    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public String getCode() {
        return code;
    }

    public void setCode(String code) {
        this.code = code;
    }

    public ProjectStatus getStatus() {
        return status;
    }

    public void setStatus(ProjectStatus status) {
        this.status = status;
    }

    public ProjectPriority getPriority() {
        return priority;
    }

    public void setPriority(ProjectPriority priority) {
        this.priority = priority;
    }

    public LocalDate getStartDate() {
        return startDate;
    }

    public void setStartDate(LocalDate startDate) {
        this.startDate = startDate;
    }

    public LocalDate getEndDate() {
        return endDate;
    }

    public void setEndDate(LocalDate endDate) {
        this.endDate = endDate;
    }

    public Integer getEstimatedHours() {
        return estimatedHours;
    }

    public void setEstimatedHours(Integer estimatedHours) {
        this.estimatedHours = estimatedHours;
    }

    public Integer getActualHours() {
        return actualHours;
    }

    public void setActualHours(Integer actualHours) {
        this.actualHours = actualHours;
    }

    public Double getBudget() {
        return budget;
    }

    public void setBudget(Double budget) {
        this.budget = budget;
    }

    public String getCurrency() {
        return currency;
    }

    public void setCurrency(String currency) {
        this.currency = currency;
    }

    public Boolean getIsPublic() {
        return isPublic;
    }

    public void setIsPublic(Boolean isPublic) {
        this.isPublic = isPublic;
    }

    public String getTags() {
        return tags;
    }

    public void setTags(String tags) {
        this.tags = tags;
    }

    public String getSettings() {
        return settings;
    }

    public void setSettings(String settings) {
        this.settings = settings;
    }

    public User getCreator() {
        return creator;
    }

    public void setCreator(User creator) {
        this.creator = creator;
    }

    public List<Task> getTasks() {
        return tasks;
    }

    public void setTasks(List<Task> tasks) {
        this.tasks = tasks;
    }

    public List<ProjectMember> getMembers() {
        return members;
    }

    public void setMembers(List<ProjectMember> members) {
        this.members = members;
    }

    public List<FileAttachment> getAttachments() {
        return attachments;
    }

    public void setAttachments(List<FileAttachment> attachments) {
        this.attachments = attachments;
    }
}

/**
 * Project status enumeration
 */
enum ProjectStatus {
    PLANNING,
    ACTIVE,
    ON_HOLD,
    COMPLETED,
    CANCELLED,
    ARCHIVED
}

/**
 * Project priority enumeration
 */
enum ProjectPriority {
    LOW,
    MEDIUM,
    HIGH,
    CRITICAL
}

File: frontend/src/types/project.ts
export interface Project {
  id: string;
  name: string;
  description: string;
  code: string;
  status: 'PLANNING' | 'IN_PROGRESS' | 'ON_HOLD' | 'COMPLETED' | 'CANCELLED';
  priority: 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';
  startDate?: string;
  endDate?: string;
  estimatedHours?: number;
  budget?: number;
  currency?: string;
  isPublic: boolean;
  tags: string[];
  createdAt: string;
  updatedAt: string;
  tenantId: number;
}

export interface ProjectRequest {
  name: string;
  description: string;
  code: string;
  status: Project['status'];
  priority: Project['priority'];
  startDate?: string;
  endDate?: string;
  estimatedHours?: number;
  budget?: number;
  currency?: string;
  isPublic: boolean;
  tags: string[];
}

export interface Task {
  id: string;
  title: string;
  description: string;
  taskNumber: string;
  status: 'TODO' | 'IN_PROGRESS' | 'DONE';
  priority: 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';
  type: 'FEATURE' | 'BUG' | 'IMPROVEMENT' | 'DOCUMENTATION';
  projectId: string;
  assignedToId?: string;
  dueDate?: string;
  startDate?: string;
  estimatedHours?: number;
  storyPoints?: number;
  tags: string[];
  createdAt: string;
  updatedAt: string;
  tenantId: number;
}

export interface TaskRequest {
  title: string;
  description: string;
  taskNumber: string;
  status: Task['status'];
  priority: Task['priority'];
  type: Task['type'];
  projectId: string;
  assignedToId?: string;
  dueDate?: string;
  startDate?: string;
  estimatedHours?: number;
  storyPoints?: number;
  tags: string[];
}

export interface PaginatedResponse<T> {
  content: T[];
  totalElements: number;
  totalPages: number;
  size: number;
  number: number;
  first: boolean;
  last: boolean;
}
